File: moseoh/_9주차/뒤에있는큰수찾기/java/Solution.java
package _9주차.뒤에있는큰수찾기.java;

import java.util.Arrays;
import java.util.Stack;

class Solution {
    public static void main(String[] args) {
        int[] numbers1 = new int[]{2, 3, 3, 5};
        int[] answer1 = new int[]{3, 5, 5, -1};
        int[] result1 = new Solution().solution(numbers1);
        PRINT_RESULT(1, result1, answer1);

        int[] numbers2 = new int[]{9, 1, 5, 3, 6, 2};
        int[] answer2 = new int[]{-1, 5, 6, 6, -1, -1};
        int[] result2 = new Solution().solution(numbers2);
        PRINT_RESULT(2, result2, answer2);
    }

    public static void PRINT_RESULT(int index, int[] result, int[] answer) {
        boolean correct = Arrays.equals(result, answer);
        StringBuilder sb = new StringBuilder();
        sb.append("\n\n테스트 케이스 ").append(index).append(": ");
        sb.append(correct ? "정답" : "오답").append("\n");
        sb.append("\t- 실행 결과: \t").append(Arrays.toString(result)).append("\n");
        sb.append("\t- 기댓값: \t").append(Arrays.toString(answer)).append("\n");
        if (correct) System.out.println(sb);
        else throw new RuntimeException(sb.toString());
    }

    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);
        Stack<Integer> stack = new Stack<>();
        for (int i = answer.length - 1; i >= 0; i--) {
            while (!stack.isEmpty() && stack.peek() <= numbers[i]) {
                stack.pop();
            }
            if (!stack.isEmpty()) {
                answer[i] = stack.peek();
            }
            stack.push(numbers[i]);
        }

        return answer;
    }
}

File: moseoh/_9주차/뒤에있는큰수찾기/kotlin/Solution.kt
package _9주차.뒤에있는큰수찾기.kotlin

import java.util.*

fun main() {
    fun printResult(index: Int, result: IntArray, answer: IntArray) {
        val correct = result.contentEquals(answer)
        val sb = StringBuilder()
        sb.append("\n\n테스트 케이스 ").append(index).append(": ")
        sb.append(if (correct) "정답" else "오답").append("\n")
        sb.append("\t- 실행 결과: \t").append(result.contentToString()).append("\n")
        sb.append("\t- 기댓값: \t").append(answer.contentToString()).append("\n")
        if (correct) println(sb) else throw RuntimeException(sb.toString())
    }

    val numbers1 = intArrayOf(2, 3, 3, 5)
    val answer1 = intArrayOf(3, 5, 5, -1)
    val result1 = Solution().solution(numbers1)
    printResult(1, result1, answer1)

    val numbers2 = intArrayOf(9, 1, 5, 3, 6, 2)
    val answer2 = intArrayOf(-1, 5, 6, 6, -1, -1)
    val result2 = Solution().solution(numbers2)
    printResult(2, result2, answer2)
}

class Solution {
    fun solution(numbers: IntArray): IntArray {
        val answer = IntArray(numbers.size) { -1 }
        val stack = Stack<Int>()
        for (i in answer.size - 1 downTo 0) {
            while (stack.isNotEmpty() && stack.peek() <= numbers[i]) {
                stack.pop()
            }
            if (stack.isNotEmpty()) {
                answer[i] = stack.peek()
            }
            stack.push(numbers[i])
        }
        return answer
    }
}